### DISCUSSION TITLE
NodeJS connection pooling and PlanetScale

### DISCUSSION THREAD

**Discussion by davidmcl** [planetscale/discussion#188] [Q&A] [answered]
Per your docs you mention `Built-in connection pooling means you’ll never run into connection limits for your database.`

... and your sample code shows 
```
const mysql = require('mysql2');
const connection = mysql.createConnection(process.env.DATABASE_URL);

connection.connect();
connection.query(.....
```

**Question:**

I already use connectionPooling in my API via `mysql2/promise createPool` 

Going forward with PlanetScale, am I shooting myself in the foot using createPool in the API with a connection limit I specify? 
Should I drop the pooling in my API altogether and just use createConnection because PlanetScale will automatically pool? 
Or does built-in pooling NOT equate to automatic pooling?
Should I still use createPool, but specify a ridiculously high connection limit, since the default is 10 if not specified ?  :-)

A little guidance please...

Thanks,
David

**Comment by jgreet** [comment, marked as answer]
There's no harm in using a connection pool on the application side and there's still something to be gained by reusing connections in terms of overhead and initial connection time. 

We recommend setting your connection pool size to the expected concurrent connection count for each instance of your application and relying on PlanetScale's connection pooling to prevent the database from being overwhelmed if you need to scale up the number of application instances.

**Reply by jcohenho** [reply]
@jgreet I'm using Prisma, Vercel and Planetscale and running into connection timeout errors: 
```Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)```

This is following a bunch of parallel requests like:
```
Promise.all() {
  query1,
  query2,
  query3
  query4,
  ...
}
```

I don't have a connection limit set and it looks like it's using the default of `5` according to the error above. With regards to your recommendation: 
> We recommend setting your connection pool size to the expected concurrent connection count for each instance of your application

The expected concurrent connection count is dynamic for my application and the maximum number of concurrent executions for [Serverless Functions on Vercel](https://vercel.com/docs/concepts/limits/overview#serverless-function-concurrency) is `1000`. Should set it to something high like that? Or should I do the opposite and set the connection_limit to `1` as [suggested by](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#optimizing-the-connection-pool) Prisma?

**Reply by jgreet** [reply]
The way I understand it is that you're creating one connection pool per instance of PrismaClient. So you're not setting a total pool size for all of the potential `1000` concurrent lambdas but for each individual lamda, since ideally they would have one PrismaClient each.

If you set `connection_limit` to `1` all of the queries will run serially. If you didn't also increase the pool timeout I imagine you would see more timeouts as queries pass 10 seconds waiting in the queue.

Prisma's advice here sounds good to me. Start at 1, or 5 since you're there already, and tune upwards
https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#optimizing-for-parallel-requests
>Tune the connection_limit parameter until a traffic spike:
>
>    Does not exhaust the database connection limit
>    Does not result in pool timeout errors

If you can handle the wait for these particular queries you could also try upping the pool time out.

I'll also add that it might be worth looking for queries that you could speed up so that connections are returned to the pool faster. [Query Insights](https://planetscale.com/docs/concepts/query-insights) might be useful there.

If you need a general rundown of connection pooling on Prisma I found this talk to be a great supplement to their documentation: https://www.youtube.com/watch?v=SmKJnITMQZw

**Reply by jcohenho** [reply]
Wonderful summary, and thank you for sharing the video resource @jgreet!

**Reply by janpio** [reply]
Prisma here, just to confirm that what @jgreet recommended here is great. 

```
Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)
```

The error message you are getting indicates that your app tries to execute more queries in parallel than what your current connection pool can handle. If the limit is 5, it can only run 5 queries in parallel. If each of those queries take 2 second, then the timeout will for sure be triggered if you start more than 5*2*10=25 queries in parallel. (Often earlier because of connection opening time for each of the connections in the connection pool).

Solutions can be to increase the pool size (5=>?), decrease the query duration (ask for less data, or make the query faster somehow with indexes and such), or increase the timeout (10=>?, if you are actually fine waiting a bit longer). If your serverless function and database are not in the same region yet, fixing that is also a great idea (as it will make both connection creation and queries faster). (Note that just increasing the pool size will often not lead to better results, as each connection needs to be created first - which is often the slowest part - so it might be better to only open x connections, and then let them be run serially in part. Balancing act.)

With Planetscale as your database, you are probably safe to create a rather high pool size as even with many parallel function invocations, PS can usually handle the many open connections.

We also have a commercial product Data Proxy that helps by moving the connection pooling to a server instead of doing it in the serverless functions: https://www.prisma.io/data-platform/proxy It works great with PlanetScale databases.

PlanetScale itself also has a serverless database client that helps with the problem, but that one is unfortunately not usable with Prisma yet (and will not be for some while).

**Reply by dani-mp** [reply]
So if I understand correctly, for an app running serverless with PlanetScale and Prisma, it seems the best solution is to increase the connection pool size considerably.

PlanetScale can handle it, and because the Prisma client can potentially be reused between lambdas (warm start), and we don't have control over this, a big connection pool size would allow more concurrent connections on that reused client.

Lambdas that create their clients (cold start) will also have their pool, and PlanetScale will still be able to handle them.

Having a connection pool size of 1 for all clients would make all lambdas run queries serially for each reused client, which seems potentially slower than the penalty we would pay for creating new connections because of the big pool.

Is this analysis correct?

**Reply by janpio** [reply]
Fundamentally, yes.
Not that if you "only" have 2 queries, it might be faster to run them sequentially instead of opening a second connection in some cases depending on when that query is triggered. But that case is pretty special.

**Reply by joelhooks** [reply]
@dani-mp what did you ultimately do here? I'm curious what "considerably" means in context.

**Reply by dani-mp** [reply]
@joelhooks I started using a third party for [enqueued operations](https://vercel.com/guides/how-to-setup-cron-jobs-on-vercel#providers) (cron jobs, bulk requests, retries., etc.; I chose Zeplo) and all my problems disappeared.

When I have a lot of concurrent traffic, Vercel throttling requests is currently limiting me more than PlanetScale or Prisma.

**Comment by christianlovescode** [comment]
Adding on here for anyone else who finds this thread – the suggestion by @janpio to use [Prisma's Data Proxy]( https://www.prisma.io/data-platform/proxy) worked wonders for me. 

Make sure to add the --data-proxy flag to your predev command in your package.json and boom, you're off to the races.

Thanks for the suggestion!

docs [here](https://www.prisma.io/docs/data-platform/data-proxy/use-data-proxy)
